smart/mesh_tools.py

```diff
@@ -812,6 +826,161 @@ def meshSizeCallback(dim, tag, x, y, z, lc):
     return (dmesh, mf1, mf2)
 
 
+def compute_curvature(
+    ref_mesh: d.mesh,
+    mf_facet: d.MeshFunction,
+    mf_cell: d.MeshFunction,
+    facet_marker_vec: list,
+    cell_marker_vec: list,
+    half_mesh_data: Tuple[d.Mesh, d.MeshFunction] = "",
+    axisymm: bool = False,
+):
+    """
+    Use dolfin functions to estimate curvature on boundary mesh.
+    Boundary meshes are created by extracting the Meshview associated
+    with each facet marker value in facet_marker_vec.
+    The length of cell_marker_vec must be the same length as facet_marker_vec,
+    with each value in the list identifying the marker value for a domain that
+    contains the boundary identified by the associated facet_marker.
+    For instance, if facet_marker_vec=[10,12] and cell_marker_vec=[1,2], then
+    the domain over which mf_cell=1 must contain the domain mf_facet=10 on its boundary
+    and the domain mf_cell=2 must contain domain mf_facet=12 on its boundary.
+    Mean curvature is approximated by first projecting facet normals (n) onto a boundary
+    finite element space and then solving a variational form of kappa = -div(n).
+
+    Args:
+        ref_mesh: dolfin mesh describing the entire geometry (parent mesh)
+        mf_facet: facet mesh function with boundary domain markers
+        mf_cell: cell mesh function with cell domain markers
+        facet_marker_vec: list with values of facet markers to iterate over
+        cell_marker_vec: list with values of cell markers (see above)
+        half_mesh_data: tuple with dolfin mesh for half domain and
+                        facet mesh function over half domain. If not specified,
+                        this is empty and the curvature values are not mapped
+                        onto the half domain
+    Returns:
+        kappa_mf: Vertex mesh function containing mean curvature values
+    """
+
+    if len(facet_marker_vec) != len(cell_marker_vec):
+        ValueError("Each listed facet marker value must have an associated cell marker value")
+
+    if half_mesh_data == "":
+        kappa_mf = d.MeshFunction("double", ref_mesh, 0)
+    else:
+        (dmesh_half, mf2_half) = half_mesh_data
+        kappa_mf = d.MeshFunction("double", dmesh_half, 0)
+
+    for i in range(len(cell_marker_vec)):
+        mesh = d.MeshView.create(mf_cell, cell_marker_vec[i])
```

Comment:
Note for later: we could use the new `create_meshview` function in dolfin to avoid some, But not all parallel issued